### PR TITLE
Mq improvements

### PR DESCRIPTION
**Pull request by TomKaltz**
A number of fixes to mq interface and implementation see commit history

### CODE REVIEW

**Comment by TomKaltz** [comment]
I realized the locked_until acts a sort of retry delay, keeping the same message from being received again for a certain timeout even if the row lock is released.  I think that sort of functionality should be up to the adapter/broker whether it applies a delay before retrying a failed message from a stream.  Let me know what you think @Rotorsoft

File: libs/eventually-pg/src/seed.ts [hunk 1]
```diff
@@ -172,6 +172,6 @@ CREATE TABLE IF NOT EXISTS public."${table}"
   data jsonb NOT NULL,
   created timestamptz NOT NULL DEFAULT now(),
   locked_until timestamptz,
-  CONSTRAINT message_queue_unique_stream_id UNIQUE (stream, id)
+  CONSTRAINT "${table}_queue_unique_stream_id" UNIQUE (stream, id)
```
Comment (Rotorsoft) [inline libs/eventually-pg/src/seed.ts:175]: the idea here is to name the index with a distinct name for each queue table

File: libs/eventually-pg/src/__tests__/message-queue.spec.ts [hunk 2]
```diff
@@ -68,4 +68,135 @@ describe("message queue", () => {
     );
     expect(messages.length).toBe(3);
   });
+
```
Comment (Rotorsoft) [inline libs/eventually-pg/src/__tests__/message-queue.spec.ts:71]: Nice 👍

File: libs/eventually-pg/src/PostgresMessageQueue.ts [hunk 3]
```diff
@@ -84,34 +81,22 @@ export const PostgresMessageQueue = <M extends Messages>(
...
           await client.query("ROLLBACK");
-          return;
+          return false;
         }
 
-        // update lock to prevent other consumers from accessing it
-        const message = next[0];
-        await client.query(
-          `
-          UPDATE "${table}" SET locked_until = NOW() + INTERVAL '1 millisecond' * $1
```
Comment (Rotorsoft) [inline libs/eventually-pg/src/PostgresMessageQueue.ts:94]: trying to find out where is the lock/timeout set before processing

Comment (TomKaltz) [reply libs/eventually-pg/src/PostgresMessageQueue.ts:94]: before the changes I made, that update would get committed if the handler failed.  I vote that we do not implement this timeout and just rollback then return the error to the caller so the caller can decide to retry immediately, retry with delay, circuit break, etc.

**Comment by Rotorsoft** [comment]
> I realized the locked_until acts a sort of retry delay, keeping the same message from being received again for a certain timeout even if the row lock is released. I think that sort of functionality should be up to the adapter/broker whether it applies a delay before retrying a failed message from a stream. Let me know what you think @Rotorsoft

The idea of a MQ is to serialize messages to the consumers, so you cannot skip locked messages. My goal is to replace most of the broker with MQs per consumer.

**Comment by TomKaltz** [comment]
> > I realized the locked_until acts a sort of retry delay, keeping the same message from being received again for a certain timeout even if the row lock is released. I think that sort of functionality should be up to the adapter/broker whether it applies a delay before retrying a failed message from a stream. Let me know what you think @Rotorsoft
> 
> The idea of a MQ is to serialize messages to the consumers, so you cannot skip locked messages. My goal is to replace most of the broker with MQs per consumer.

I'm implementing a POC processor/broker right now that tries to sequentially consume multiple streams (up to configurable max slots) in parallel from one of these queues.  Currently I'm using the following query to discover consumable streams ...
```
WITH available_messages AS (
            SELECT id, stream
            FROM "${options.tableName}"
            WHERE locked_until IS NULL OR locked_until <= NOW()
            FOR SHARE SKIP LOCKED
          )
          SELECT DISTINCT stream
          FROM available_messages
          LIMIT $1
```

the implementation is currently naive and that query does return stream names that could already be processing but I'm handling this concurrency elsewhere in the code.  I'll let you know if I ever get it working fully.

Regarding guarding for consumer/stream concurrency at the PostgresMessageQueue level, I would love that but we might have to store (stream, consumer) locks in another table to make this work correctly.  That would allow for distributed processing (ultimate dream) but for now I'm just experimenting with single-process in-memory concurrency control for consuming multiple streams in parallel.

File: libs/eventually-pg/src/PostgresMessageQueue.ts (hunk 3 above)
Comment (Rotorsoft) [reply libs/eventually-pg/src/PostgresMessageQueue.ts:94]: ok... this works but you need a broker to coordinate retries or put that responsibility on the consumers... part of this exercise is to determine what works best, push or pull model.

Comment (Rotorsoft) [reply libs/eventually-pg/src/PostgresMessageQueue.ts:94]: we probably don't need the timeout param in the interface anymore

**Comment by Rotorsoft** [comment]
yep, this PG adapter is a little naive, with one central queue per consumer group competing for messages.... this can evolve into distributed queue adapters (redis, kafka, rmq) in real apps

**Comment by TomKaltz** [comment]
> yep, this PG adapter is a little naive, with one central queue per consumer group competing for messages.... this can evolve into distributed queue adapters (redis, kafka, rmq) in real apps

In researching the topic of distributed stream processing it seems that the virtual actor pattern in the actor model solves the distributed processing issue but just like with event sourcing, that aren't a lot of options for this in typescript.  [Dapr](https://docs.dapr.io/developing-applications/building-blocks/actors/actors-overview/) seems like the closest option and is compatible with javascript but it might be a little complex.

I've become increasingly more interested in NATS for inter-process communication and it's Jetstream product.  It seems like that technology could be used to develop a lighter virtual actor implementation.  Alas, that sort of experimentation is for another day.

**Review by Rotorsoft — APPROVED** [review]
LGTM 🚀

**Pull request merged**